### PR TITLE
couch to sql migration part 2

### PR DESCRIPTION
**Pull request by czue**
@dimagi/scale-team had some time to work a little more on this today. I think it's basically working (though obviously lots of edge cases and testing to do). we also still need to do the case update audit.

cc @emord

### CODE REVIEW

File: corehq/form_processor/management/commands/migrate_domain_from_couch_to_sql.py [hunk 1]
```diff
@@ -54,6 +55,8 @@ def _get_main_form_iterator(domain):
 
 def _migrate_form_and_associated_models(domain, couch_form):
     sql_form = _migrate_form_and_attachments(domain, couch_form)
+    # todo: this should hopefully not be necessary once all attachments are in blobDB
+    sql_form.get_xml = MagicMock(return_value=couch_form.get_xml())
```
Comment (benrudolph) [inline corehq/form_processor/management/commands/migrate_domain_from_couch_to_sql.py]: 🎱

File: corehq/form_processor/tests/test_migration.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        FormProcessorTestUtils.delete_all_cases_forms_ledgers()
+        self.domain = uuid.uuid4().hex
+
+    def test_basic_form_migration(self):
+        create_and_save_a_form(self.domain)
+        self.assertFalse(should_use_sql_backend(self.domain))
+        self.assertEqual(1, len(FormAccessors(domain=self.domain).get_all_form_ids_in_domain()))
+        self._do_migration_and_assert_flags(self.domain)
+        self.assertEqual(1, len(FormAccessors(domain=self.domain).get_all_form_ids_in_domain()))
+        # todo: verify form properties?
```
Comment (benrudolph) [inline corehq/form_processor/tests/test_migration.py]: seems like we should have at least one test do a diff of the form json just for sanity

**Comment by benrudolph** [comment]
not sure if related, but: 

```
======================================================================
FAIL: test_case_close_format (corehq.apps.repeaters.tests.test_repeater.CaseRepeaterTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 232, in inner
    return helper(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 205, in __call__
    call_with_settings(fn_with_pre_and_post, run_config.settings, args, kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 220, in call_with_settings
    fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 201, in fn_with_pre_and_post
    self.fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/repeaters/tests/test_repeater.py", line 275, in test_case_close_format
    CaseFactory().close_case(CASE_ID)
  File "corehq/ex-submodules/casexml/apps/case/mock.py", line 320, in close_case
    return self.create_or_update_case(CaseStructure(case_id=case_id, attrs={'close': True}))[0]
  File "corehq/ex-submodules/casexml/apps/case/mock.py", line 323, in create_or_update_case
    return self.create_or_update_cases([case_structure], form_extras)
  File "corehq/ex-submodules/casexml/apps/case/mock.py", line 346, in create_or_update_cases
    return list(CaseAccessors(self.domain).get_cases(case_ids, ordered=True))
  File "/mnt/commcare-hq-ro/corehq/form_processor/interfaces/dbaccessors.py", line 274, in get_cases
    return self.db_accessor.get_cases(case_ids, ordered=ordered)
  File "submodules/dimagi-utils-src/dimagi/utils/decorators/memoized.py", line 11, in _inner
    return m(*args, **kwargs)
  File "submodules/dimagi-utils-src/dimagi/utils/decorators/memoized.py", line 128, in __call__
    value = self.func( *args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/form_processor/interfaces/dbaccessors.py", line 265, in db_accessor
    if should_use_sql_backend(self.domain):
  File "/mnt/commcare-hq-ro/corehq/form_processor/utils/general.py", line 24, in should_use_sql_backend
    domain_name, domain_object = _get_domain_name_and_object(domain_object_or_name)
  File "/mnt/commcare-hq-ro/corehq/form_processor/utils/general.py", line 51, in _get_domain_name_and_object
    return domain_object_or_name, Domain.get_by_name(domain_object_or_name)
  File "/mnt/commcare-hq-ro/corehq/util/quickcache.py", line 312, in inner
    return helper(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/quickcache.py", line 201, in __call__
    return super(SkippableQuickCache, self).__call__(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/quickcache.py", line 100, in __call__
    content = self.fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/domain/models.py", line 537, in get_by_name
    _assert(False, '%r is not a valid domain name' % name)
  File "/mnt/commcare-hq-ro/corehq/util/soft_assert/core.py", line 33, in __call__
    return self._call(assertion, msg, obj)
  File "/mnt/commcare-hq-ro/corehq/util/soft_assert/core.py", line 47, in _call
    raise AssertionError(msg)
AssertionError: None is not a valid domain name
```

File: corehq/form_processor/tests/test_migration.py (hunk 2 above)
Comment (czue) [inline corehq/form_processor/tests/test_migration.py]: hard to argue with that.

**Comment by czue** [comment]
:white_check_mark: (will expand tests later but dropping for interrupt stuff for a bit)

**Pull request merged**